lib/embulk/input/service/jql_service.rb

```diff
@@ -0,0 +1,167 @@
+require 'embulk/input/service/base_service'
+
+module Embulk
+  module Input
+    module Service
+      class JqlService < BaseService
+
+        def validate_config
+          super
+          jql_script = @config.param(:jql_script, :string, default: nil)
+
+          validate_jql_script(jql_script)
+        end
+
+        def create_task
+          {
+            timezone: @config.param(:timezone, :string, default: ""),
+            api_secret: @config.param(:api_secret, :string),
+            export_endpoint: export_endpoint,
+            incremental: @config.param(:incremental, :bool, default: true),
+            dates: range,
+            schema: @config.param(:columns, :array),
+            retry_initial_wait_sec: @config.param(:retry_initial_wait_sec, :integer, default: 1),
+            retry_limit: @config.param(:retry_limit, :integer, default: 5),
+            jql_mode: true,
+            jql_script: @config.param(:jql_script, :string, nil)
+          }
+        end
+
+        def create_next_config_diff(task_report)
+          next_to_date = Date.parse(task_report[:to_date]).next_day(1)
+          {
+            from_date: next_to_date.to_s
+          }
+        end
+
+        def guess_columns
+          range = guess_range
+          # giveup_when_mixpanel_is_down
+          Embulk.logger.info "Guessing schema using #{range.first}..#{range.last}"
+          client = create_client
+
+          sample_records = client.send_jql_script_small_dataset(parameters(range.first, range.last, @config.param(:jql_script, :string, nil)))
+          guess_from_records(sample_records)
+        end
+
+        def ingest(task, page_builder)
+          giveup_when_mixpanel_is_down
+          @dates = task[:dates]
+          @schema = task[:schema]
+          @timezone = task[:timezone]
+
+          client = create_client
+
+          if preview?
+            sample_records = client.send_jql_script_small_dataset(parameters(@dates.first, @dates.last, task[:jql_script]))
```

Comment:
I'm not sure why slicing the dates when we only use the first and last dates? Is it possible to apply the same slicing strategy (per days, I believed) here for JQL? If not, then how well this will handle large datasets, looks like it brings the entire result into memory all at once?

Reply:
I think slicing the dates help to get the data faster. We don't really need full data in guess and preview. 
The maximum dataset that this endpoint can return is 2 GB, mentioned here.
https://developer.mixpanel.com/docs/jql-getting-started#section-running-saving-queries-within-mixpanel
 
One day slicing could bring the rate limit problem. The rate limit of this endpoint is 60 requests in 1 hour but it doesn't provide waiting time for the next request. 
Should we provide sliding days option here? Default is something like 7 days?

Reply:
Sorry that I may put the comment on a confusing line, the question is also meant for the actual run, where it also uses a one-shot request. 
I may misunderstand or something, I didn't mean we should adjust the slice size to 1-day (now I read the existing export API more carefully, it's 7-days slice), I meant I couldn't find where we slicing or partitioning the requests at all.

It's more about the concern with memory footprint rather than for the sake of concurrency/job time. Existing implementation for export API seems to do sequential requests with (by default) 7-days slices, and it's going fine so far I supposed?

Reply:
Discussed offline and we aligned.